Blake2

waywardgeek opened the pull request:
This set of changes adds initial BLAKE2 support to OpenSSL.  The BLAKE2s is a 32-bit friendly 256-bit hash, while BLAKE2b is a 64-bit friendly 512-bit hash.  The point of BLAKE2 is extremely fast secure hashing, even faster than MD5 on modern desktop CPUs with SSE instructions, as well as on ARM NEON processors.

This PR integrates, with the authors' permission, the portable reference code from blake2.net, licensed under the OpenSSL license.  EVP support is included, and test vectors for BLAKE2b and BLAKE2s have been added to tests/evptests.txt.

richsalz commented:
Way cool.

It would be good to integrate into the dgst or enc program?
Doc updates?

waywardgeek commented:
Sounds good to me.  I'll have to figure out what the enc and dgst programs are, but I'm on it :)

waywardgeek commented:
I added what I _think_ is needed to the docs, and support for blake2b and blake2s in the dgst command.  What does it means to integrate it into the "enc" command?  Also, I did not add the RSA versions of the blake2 functions.  Is there a use case?

kroeckx commented:
So at least travis is returning errors like: "Error: EVP_blake2b does not have a number assigned"

kroeckx commented:
I understand that you're not actually the author of the code, but you say that it's with his permission.  Do you have some reference to this permission?  Maybe we should use the CLA for this?

richsalz commented:
The original code is in the public domain.

kroeckx commented on crypto/blake2/blake2b.c:
```diff
@@ -0,0 +1,376 @@
+    memset(block, 0, BLAKE2B_BLOCKBYTES);
+    memcpy(block, key, keylen);
+    BLAKE2b_Update(c, block, BLAKE2B_BLOCKBYTES);
+    secure_zero_memory(block, BLAKE2B_BLOCKBYTES); /* Burn the key from stack */
```
secure_zero_memory() should probably be replaced with OPENSSL_cleanse()

kroeckx commented:
So, looking at the code, it actually has various license in it, it should be fine.

kroeckx commented on include/openssl/blake2.h:
```diff
@@ -0,0 +1,57 @@
+unsigned char *BLAKE2s(const unsigned char *data, size_t datalen,
+                       const unsigned char *key, size_t keylen,
+                       unsigned char *md);
+
```
I think we would just like to see an EVP interface, this whole header and the documentation related to it should probably be removed.

richsalz replied:
I agree, only expose the EVP.

waywardgeek commented:
I removed blake2.h so that only the EVP API is exposed, and switched to using OPENSSL_cleanse.  I fixed the failing test case by adding blake2 to the list of known algorithms in mkdef.pl.  Are there any additional issues?  Thanks.

richsalz commented:
I don't have any at this time, but we'll see what other reviewers think :)

kroeckx commented:
You should probably update apps/openssl.c's list_disabled() function.

richsalz commented:
So if you fix @kroeckx 's issue, and someone else on the team reviews it, we can hopefully put this into the next alpha release.
+1

waywardgeek commented:
I did as @droechkx suggested and added blake2 to the list of disables.  This should work fine.  However, I tried doing a ./config --no-blake2, and the compile fails.  It also fails for --no-aes, with the message:

gcc: error: unrecognized command line option '--no-aes'

I think this has to do with the recent restructuring of the build system.  I verified that master has this problem, too.

richsalz commented:
Use a single dash.

waywardgeek commented:
That worked.  I fixed a couple goobers that crept in due to the build system revamp, and now the -no-blake2 flag works, and blake2b and blake2s show up in apps/openssl help.  In theory, this is ready to merge, at least if there are no new conflicts in master that were checked in today.

richsalz commented:
I'm okay with this.  @kroeckx , your view?  (I can merge if you want)

kroeckx commented:
The unified build option doesn't work, it doesn't seem to build all the files resulting in:

    ./libcrypto.a(c_alld.o): In function `openssl_add_all_digests_internal':
    crypto/evp/c_alld.c:(.text+0xf5): undefined reference to `EVP_blake2b'
    crypto/evp/c_alld.c:(.text+0x102): undefined reference to `EVP_blake2s'

kroeckx commented:
It has some lines with a space at the end of the line in the the copyright part.

kroeckx commented on Makefile.in:
```diff
@@ -270,6 +271,7 @@ BUILDENV=	LC_ALL=C PLATFORM='$(PLATFORM)' PROCESSOR='$(PROCESSOR)'\
 		SHA1_ASM_OBJ='$(SHA1_ASM_OBJ)'			\
 		MD5_ASM_OBJ='$(MD5_ASM_OBJ)'			\
 		RMD160_ASM_OBJ='$(RMD160_ASM_OBJ)'		\
+                BLAKE2_OBJ='$(BLAKE2_OBJ)'                      \
```
This is with spaces while all other lines have tabs there.

kroeckx commented on crypto/blake2/TODO:
```diff
@@ -0,0 +1,4 @@
+The code uses NATIVE_LITTLE_ENDIAN in #if statements.  Is this right?
```
I don't think we have any such define.  But we have L_ENDIAN and B_ENDIAN.  They might not be defined, but if they're defined you can use it optimize things knowing the endianness.  I guess you just have to replace all your cases with `#ifdef L_ENDIAN`

richsalz replied:
Yes, L_ENDIAN is defined in CFLAGS.  But also yes, prefer to do order-neutral code first.

kroeckx replied:
The code he has is either endian neutral or little endian.

levitte commented on crypto/blake2/build.info:
```diff
@@ -0,0 +1,3 @@
+LIBS=../../libcrypto
+SOURCE[../../libcrypto]=\
+        blake2b.c blake2s.c {- $target{blake2b_obj} -}
```
Correct the two file names, and `blake2b_obj` should be `blake2b_asm_src`

waywardgeek replied:
I changed $target{blake2b_obj} to $target{blake2_obj}.  Is there another filename that is wrong?  The name blake2_obj looks correct to me for projects that do not yet include assembly code files, which will change as soon as I am allowed to check in the SSE optimized versions down the road.  Maybe I don't understand the _asm_ variables.

levitte replied:
`$target{blak2_obj}` is fine for the older `Makefile.in`. The new `build.info` files deal with source, and therefore use the names from `Configurations/00-base-templates.conf` directly. That's where the asm variable comes from. 

I noticed you removed it entirely, which is technically fine now, but will not be fine as soon as asm source is added.

waywardgeek replied:
I realize I will need to do a bit of work (modify Configure, build.info,
Makefile.in, maybe a couple more), but I prefer to keep this initial PR
simple.  Thanks for reviewing it carefully!  I do make mistakes, and I feel
a lot better now that a few people have taken a good look.

On Sun, Feb 28, 2016 at 11:30 PM, Richard Levitte <notifications@github.com>
wrote:

> In crypto/blake2/build.info
> <https://github.com/openssl/openssl/pull/566#discussion_r54375731>:
>
> > @@ -0,0 +1,3 @@
> > +LIBS=../../libcrypto
> > +SOURCE[../../libcrypto]=\
> > +        blake2b.c blake2s.c {- $target{blake2b_obj} -}
>
> $target{blak2_obj} is fine for the older Makefile.in. The new build.info
> files deal with source, and therefore use the names from
> Configurations/00-base-templates.conf directly. That's where the asm
> variable comes from.
>
> I noticed you removed it entirely, which is technically fine now, but will
> not be fine as soon as asm source is added.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openssl/openssl/pull/566/files#r54375731>.
>

richsalz replied:
That’s the point of code reviews :)  Yes, assembly is harder to integrate.  But do we think we're done now, from your viewpoint?

waywardgeek replied:
Yes, I think we're done for now.  I do want to revisit this in the
not-too-distant future to add SSE support with some asm code, but it looks
good to me as-is.  The triple-license has also been incorporated upstream
now.

On Mon, Feb 29, 2016 at 3:43 PM, Rich Salz <notifications@github.com> wrote:

> In crypto/blake2/build.info
> <https://github.com/openssl/openssl/pull/566#discussion_r54499900>:
>
> > @@ -0,0 +1,3 @@
> > +LIBS=../../libcrypto
> > +SOURCE[../../libcrypto]=\
> > +        blake2b.c blake2s.c {- $target{blake2b_obj} -}
>
> That’s the point of code reviews :) Yes, assembly is harder to integrate.
> But do we think we're done now, from your viewpoint?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openssl/openssl/pull/566/files#r54499900>.
>

kroeckx commented on crypto/blake2/TODO:
```diff
@@ -0,0 +1,4 @@
+The code uses NATIVE_LITTLE_ENDIAN in #if statements.  Is this right?
+
+The code used #pragma pack, which I removed.  The layout for the param structs
+is required to be fully packed, but I think it will be on all platforms.
```
I'm not sure that's actually going to be the case.  Could you add a test for it checking the size of the struct?

richsalz replied:
I didn't realize the fully-packed requirement.  You're almost definitely going to have to lay out the bytes yourself.

kroeckx replied:
As far as I can see, they are actually.  I assume it's those structs that have comments with the byte offset in it.

waywardgeek commented:
I addressed all of the comments above, accept for levitte's comment where I still have a couple questions.

waywardgeek commented:
I removed the $target clause from blake2/build.info, which I think is the right action.  I copied that file from another directory that had assembly code files, which currently I have not added to blake2.

So, I think I've addressed everything.  It builds now with --unified, I got rid of the trailing white space, changed spaces to tabs where the other lines used tabs, changed the #ifdefs to use L_ENDIAN, and added an OPENSSL_assert for the size of the "param" structs to catch any errors on platforms that for some reason have unexpected structure padding.  I manually tested that the assertions catch padding errors.  Is there anything else I should do?  I left in the L_ENDIAN rather than removing them, and put in the OPENSSL_assert for the padding struct sizes because both of these are in speed critical code, and a lot of platforms will compile with SSE support disabled.  This version of BLAKE2 is reasonably fast without SSE, partly because of such optimizations.

richsalz commented:
Nice effort.  +1 from me still and I'll merge once another review.

t-j-h commented on crypto/blake2/blake2_impl.h:
```diff
@@ -0,0 +1,145 @@
+ *
+ * - OpenSSL license   : https://www.openssl.org/source/license.html
+ * - Apache 2.0        : http://www.apache.org/licenses/LICENSE-2.0
+ * - CC0 1.0 Universal : http://www.apache.org/licenses/LICENSE-2.0
```
Your CC0 link is wrong. You probably meant https://creativecommons.org/publicdomain/zero/1.0/

waywardgeek commented:
Nice catch with the license link error.  The author, Samuel Neves, has a valid complaint about the difficulty of giving software away, since he put his code as much into the public domain.  Zooko suggested the triple license, which I proposed to Samuel, who agreed.  However, the typo is mine.  I'll do a pull request to the upstream Blake2 code base to get the triple license there so we wont have to deal with this again.

richsalz commented:
+1 okay, thanks for your efforts.  will integrate after someone else on the team reviews.

waywardgeek commented:
I re-merged with master so it would build again.  Anytime util/libeay.num
changes, I get a merge conflict.  I also cleaned up the blake2
functionality again in apps/openssl.  I assume we want blake2b and blake2s
advertised in "openssl help", since "openssl blake2b" works.  For speed
testing, since only EVP APIs are supported, you have to do "openssl speed
-evp blake2b".

If this is not the functionality we want, let me know and I'll fix it.

Bill

On Mon, Feb 29, 2016 at 4:04 PM, Rich Salz <notifications@github.com> wrote:

> +1 okay, thanks for your efforts. will integrate after someone else on the
> team reviews.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openssl/openssl/pull/566#issuecomment-190457078>.
>

richsalz commented:
That’s all good.  Yes, ‘make update’ whenever you rebase to master.  Don’t worry about it I'll do it for the final merge.

richsalz commented:
so @kroeckx you happy with this?

kroeckx commented:
Probably, didn't have time to actually look at the code itself.
But I'm trying to do that soon.

kroeckx commented on crypto/blake2/blake2_impl.h:
```diff
@@ -0,0 +1,145 @@
+{
+#if defined(L_ENDIAN)
+    uint32_t w;
+    memcpy(&w, src, sizeof w);
```
The style guide says to use sizeof(w)

kroeckx commented on crypto/blake2/blake2b.c:
```diff
@@ -0,0 +1,310 @@
+            c->buflen -= BLAKE2B_BLOCKBYTES;
+            in += fill;
+            datalen -= fill;
+        } else { /* inlen <= fill */
```
That should be datalen, not inlen.

kroeckx commented on crypto/blake2/blake2b.c:
```diff
@@ -0,0 +1,310 @@
+
+    while(datalen > 0) {
+        left = c->buflen;
+        fill = 2 * BLAKE2B_BLOCKBYTES - left;
```
Can you make that fill = sizeof(c->buf) - left;?

kroeckx commented on crypto/blake2/blake2b.c:
```diff
@@ -0,0 +1,310 @@
+
+int BLAKE2b_Update(BLAKE2B_CTX *c, const void *data, size_t datalen) {
+    const uint8_t *in = data;
+    size_t left, fill;
```
I find left a confusing name here, size left is used as the current buffer length, and really just as a copy of it.

kroeckx commented on crypto/include/internal/blake2_locl.h:
```diff
@@ -0,0 +1,113 @@
+struct blake2b_ctx_st {
+    uint64_t h[8];
+    uint64_t t[2];
+    uint64_t f[2];
```
Those variable names aren't very clear.  I think "t" is counter, the (current) hash, and I'm not sure what f is about.

kroeckx replied:
f probably stands for final.

waywardgeek replied:
On Sat, Mar 5, 2016 at 10:25 AM, Kurt Roeckx <notifications@github.com>
wrote:

> In crypto/include/internal/blake2_locl.h
> <https://github.com/openssl/openssl/pull/566#discussion_r55125505>:
>
> > +    uint8_t  depth;         /* 4 */
> > +    uint32_t leaf_length;   /* 8 */
> > +    uint64_t node_offset;   /* 16 */
> > +    uint8_t  node_depth;    /* 17 */
> > +    uint8_t  inner_length;  /* 18 */
> > +    uint8_t  reserved[14];  /* 32 */
> > +    uint8_t  salt[BLAKE2B_SALTBYTES]; /* 48 */
> > +    uint8_t  personal[BLAKE2B_PERSONALBYTES];  /* 64 */
> > +};
> > +
> > +typedef struct blake2b_param_st BLAKE2B_PARAM;
> > +
> > +struct blake2b_ctx_st {
> > +    uint64_t h[8];
> > +    uint64_t t[2];
> > +    uint64_t f[2];
>
> Those variable names aren't very clear. I think "t" is counter, the
> (current) hash, and I'm not sure what f is about.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openssl/openssl/pull/566/files#r55125505>.
>
I agree.  The 'h' variable is the same as the original SHA3 BLAKE paper, so
I would prefer to leave it as-is.  I am not sure about t and f.  Generally,
I'm not commenting your suggestions that I agree with - I'm simply doing
them.  To do this one, I need more information about what t and f mean.

kroeckx replied:
I think f stands for final, where f[1] is last/final node, and f[0] is last block

t probably stands for total, it's used as a counter

waywardgeek replied:
On Sat, Mar 5, 2016 at 12:59 PM, Kurt Roeckx <notifications@github.com>
wrote:

> In crypto/include/internal/blake2_locl.h
> <https://github.com/openssl/openssl/pull/566#discussion_r55127857>:
>
> > +    uint8_t  depth;         /* 4 */
> > +    uint32_t leaf_length;   /* 8 */
> > +    uint64_t node_offset;   /* 16 */
> > +    uint8_t  node_depth;    /* 17 */
> > +    uint8_t  inner_length;  /* 18 */
> > +    uint8_t  reserved[14];  /* 32 */
> > +    uint8_t  salt[BLAKE2B_SALTBYTES]; /* 48 */
> > +    uint8_t  personal[BLAKE2B_PERSONALBYTES];  /* 64 */
> > +};
> > +
> > +typedef struct blake2b_param_st BLAKE2B_PARAM;
> > +
> > +struct blake2b_ctx_st {
> > +    uint64_t h[8];
> > +    uint64_t t[2];
> > +    uint64_t f[2];
>
> f probably stands for final.
>
In case f and t correspond to a paper the authors wrote, I would prefer to
leave these alone.  The 'h' variable was referred to as h in several places
in the paper for the SHA3 competition, so it already is understood in the
context of BLAKE/BLAKE2 by cryptanalysts.

kroeckx commented on crypto/include/internal/blake2_locl.h:
```diff
@@ -0,0 +1,113 @@
+    uint64_t f[2];
+    uint8_t  buf[2 * BLAKE2B_BLOCKBYTES];
+    size_t   buflen;
+    uint8_t  last_node;
```
last_node never seems to be set.

kroeckx commented on crypto/blake2/blake2b.c:
```diff
@@ -0,0 +1,310 @@
+            memcpy(c->buf + left, in, datalen);
+            c->buflen += datalen; /* Be lazy, do not compress */
+            in += datalen;
+            datalen -= datalen;
```
Updating in and datalen doesn't make sense here, you could just return.

kroeckx commented on crypto/include/internal/blake2_locl.h:
```diff
@@ -0,0 +1,113 @@
+    uint64_t h[8];
+    uint64_t t[2];
+    uint64_t f[2];
+    uint8_t  buf[2 * BLAKE2B_BLOCKBYTES];
```
Is there a reason it's 2 * the size?  I currently can't see the reason since it's also just compressed in Final.

kroeckx replied:
This is probably to be able to call blake2b_set_lastblock() before the last compress call, and so it should probably be documented.

waywardgeek replied:
On Sat, Mar 5, 2016 at 1:02 PM, Kurt Roeckx <notifications@github.com>
wrote:

> In crypto/include/internal/blake2_locl.h
> <https://github.com/openssl/openssl/pull/566#discussion_r55127932>:
>
> > +    uint32_t leaf_length;   /* 8 */
> > +    uint64_t node_offset;   /* 16 */
> > +    uint8_t  node_depth;    /* 17 */
> > +    uint8_t  inner_length;  /* 18 */
> > +    uint8_t  reserved[14];  /* 32 */
> > +    uint8_t  salt[BLAKE2B_SALTBYTES]; /* 48 */
> > +    uint8_t  personal[BLAKE2B_PERSONALBYTES];  /* 64 */
> > +};
> > +
> > +typedef struct blake2b_param_st BLAKE2B_PARAM;
> > +
> > +struct blake2b_ctx_st {
> > +    uint64_t h[8];
> > +    uint64_t t[2];
> > +    uint64_t f[2];
> > +    uint8_t  buf[2 * BLAKE2B_BLOCKBYTES];
>
> This is probably to be able to call blake2b_set_lastblock() before the
> last compress call, and so it should probably be documented.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openssl/openssl/pull/566/files#r55127932>.
>
This buffer is this large again to support being as lazy as possible.  The
BLAKE2 designers worked hard to make it efficient for small hashes, and I
think this is part of that effort.

kroeckx commented on crypto/blake2/blake2b.c:
```diff
@@ -0,0 +1,310 @@
+    blake2b_increment_counter(c, c->buflen);
+    blake2b_set_lastblock(c);
+    /* Padding */
+    memset(c->buf + c->buflen, 0, 2 * BLAKE2B_BLOCKBYTES - c->buflen);
```
Please use memset(c->buf + c->buflen, 0, sizeof(c->buf) - c->buflen).

I also think the "2 *" isn't needed, only the first BLAKE2B_BLOCKBYTES get used, and the rest gets cleansed later anyway.

kroeckx commented on crypto/blake2/blake2b.c:
```diff
@@ -0,0 +1,310 @@
+
+unsigned char *BLAKE2b(const unsigned char *data, size_t datalen,
+                       const unsigned char *key, size_t keylen,
+                       unsigned char *md)
```
I think this function is unused?

kroeckx commented:
Please add some minimal documentation for all the functions, including the return value.

kroeckx commented on crypto/blake2/blake2b.c:
```diff
@@ -0,0 +1,310 @@
+        store64(buffer + sizeof(c->h[i]) * i, c->h[i]);
+    }
+
+    memcpy(md, buffer, BLAKE2B_DIGEST_LENGTH);
```
Is there a reason this is first written to a local buffer and then copied to md instead of just writing it to md?

waywardgeek replied:
On Sat, Mar 5, 2016 at 10:52 AM, Kurt Roeckx <notifications@github.com>
wrote:

> In crypto/blake2/blake2b.c
> <https://github.com/openssl/openssl/pull/566#discussion_r55125892>:
>
> > +        c->buflen -= BLAKE2B_BLOCKBYTES;
> > +        memcpy(c->buf, c->buf + BLAKE2B_BLOCKBYTES, c->buflen);
> > +    }
> > +
> > +    blake2b_increment_counter(c, c->buflen);
> > +    blake2b_set_lastblock(c);
> > +    /* Padding */
> > +    memset(c->buf + c->buflen, 0, 2 * BLAKE2B_BLOCKBYTES - c->buflen);
> > +    blake2b_compress(c, c->buf);
> > +
> > +    /* Output full hash to temp buffer */
> > +    for(i = 0; i < 8; ++i) {
> > +        store64(buffer + sizeof(c->h[i]) * i, c->h[i]);
> > +    }
> > +
> > +    memcpy(md, buffer, BLAKE2B_DIGEST_LENGTH);
>
> Is there a reason this is first written to a local buffer and then copied
> to md instead of just writing it to md?
>
This deals with byte ordering.

kroeckx replied:
I don't see it.  Both are character arrays, and then you just do a memcpy() from one to the other.

waywardgeek replied:
On Tue, Mar 8, 2016 at 10:10 AM, Kurt Roeckx <notifications@github.com>
wrote:

> In crypto/blake2/blake2b.c
> <https://github.com/openssl/openssl/pull/566#discussion_r55400511>:
>
> > +        c->buflen -= BLAKE2B_BLOCKBYTES;
> > +        memcpy(c->buf, c->buf + BLAKE2B_BLOCKBYTES, c->buflen);
> > +    }
> > +
> > +    blake2b_increment_counter(c, c->buflen);
> > +    blake2b_set_lastblock(c);
> > +    /* Padding */
> > +    memset(c->buf + c->buflen, 0, 2 * BLAKE2B_BLOCKBYTES - c->buflen);
> > +    blake2b_compress(c, c->buf);
> > +
> > +    /* Output full hash to temp buffer */
> > +    for(i = 0; i < 8; ++i) {
> > +        store64(buffer + sizeof(c->h[i]) * i, c->h[i]);
> > +    }
> > +
> > +    memcpy(md, buffer, BLAKE2B_DIGEST_LENGTH);
>
> I don't see it. Both are character arrays, and then you just do a memcpy()
> from one to the other.
>
You're right.  The buffer is needed in the upstream code because the output
length is a parameter to Final, and store64 might overrun the buffer.
Fixed.

kroeckx commented on crypto/blake2/blake2b.c:
```diff
@@ -0,0 +1,310 @@
+    return blake2b_init_param(c, P);
+}
+
+int BLAKE2b_InitKey(BLAKE2B_CTX *c, const void *key, size_t keylen) {
```
This function also seems to be unused, if you remove BLAKE2b().  You might be able to remove other things too.

kroeckx commented on crypto/include/internal/blake2_locl.h:
```diff
@@ -0,0 +1,113 @@
+    BLAKE2S_KEYBYTES   = 32,
+    BLAKE2S_SALTBYTES  = 8,
+    BLAKE2S_PERSONALBYTES = 8
+};
```
I think those should be #defines instead of a enum.

kroeckx commented on crypto/blake2/blake2b.c:
```diff
@@ -0,0 +1,310 @@
+    while(datalen > 0) {
+        left = c->buflen;
+        fill = 2 * BLAKE2B_BLOCKBYTES - left;
+        if(datalen > fill) {
```
This should be >=

kroeckx replied:
So I think there are 2 options:
- Have buf be BLAKE2B_BLOCKBYTES and use > here.  It would clearly need a comment in that case.
- Have buf be 2 * BLAKE2B_BLOCKBYTES, and optionally use >= or >, with > just delaying it until the next call where you'd copy 0 bytes before compressing.

waywardgeek replied:
On Sat, Mar 5, 2016 at 12:53 PM, Kurt Roeckx <notifications@github.com>
wrote:

> In crypto/blake2/blake2b.c
> <https://github.com/openssl/openssl/pull/566#discussion_r55127766>:
>
> > +    for(i = 0; i < 8; ++i) {
> > +        S->h[i] = S->h[i] ^ v[i] ^ v[i + 8];
> > +    }
> > +
> > +#undef G
> > +#undef ROUND
> > +}
> > +
> > +int BLAKE2b_Update(BLAKE2B_CTX *c, const void *data, size_t datalen) {
> > +    const uint8_t *in = data;
> > +    size_t left, fill;
> > +
> > +    while(datalen > 0) {
> > +        left = c->buflen;
> > +        fill = 2 * BLAKE2B_BLOCKBYTES - left;
> > +        if(datalen > fill) {
>
> This should be >=
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openssl/openssl/pull/566/files#r55127766>.
>

No, > is correct.  It will actually cause the result to be incorrect if it
were changed.  BLAKE2b is as "lazy" as possible about compression.  The
last call to Update is different, so it matters whether or not this is >=
vs >.

kroeckx replied:
Yes, the last call is different, as I already explained myself.  So you want to be able to keep at least 1 block.  But with the current > and a buffer of 2 * the block size you can keep 2 blocks and you're just going to process the first one of that like normal.  And yes, you can be lazy, but I can't see that it buys you anything.  I can only see you waste something on the next Update() call.  But it's not that important.

waywardgeek replied:
On Sat, Mar 5, 2016 at 12:53 PM, Kurt Roeckx <notifications@github.com>
wrote:

> In crypto/blake2/blake2b.c
> <https://github.com/openssl/openssl/pull/566#discussion_r55127766>:
>
> > +    for(i = 0; i < 8; ++i) {
> > +        S->h[i] = S->h[i] ^ v[i] ^ v[i + 8];
> > +    }
> > +
> > +#undef G
> > +#undef ROUND
> > +}
> > +
> > +int BLAKE2b_Update(BLAKE2B_CTX *c, const void *data, size_t datalen) {
> > +    const uint8_t *in = data;
> > +    size_t left, fill;
> > +
> > +    while(datalen > 0) {
> > +        left = c->buflen;
> > +        fill = 2 * BLAKE2B_BLOCKBYTES - left;
> > +        if(datalen > fill) {
>
> This should be >=
>

I implemented your other suggestion: reducing the input buffer size by 2X.
With that change, it is important that this be > 0 so that we do not
compress the last block, which is hashed differently.

kroeckx commented on crypto/blake2/blake2b.c:
```diff
@@ -0,0 +1,310 @@
+
+int BLAKE2b_Final(unsigned char *md, BLAKE2B_CTX *c)
+{
+    uint8_t buffer[BLAKE2B_OUTBYTES] = {0};
```
This initialization seems to be pointless?

kroeckx commented:
blake2b.c and blake2s.c seem to contain mostly the same code.  Could please reuse the code using defines in the header file so that the differences between the 2 is in the .c file?  You might want to look at what md32_common.h is doing.

waywardgeek commented:
On Sat, Mar 5, 2016 at 1:11 PM, Kurt Roeckx <notifications@github.com>
wrote:

> blake2b.c and blake2s.c seem to contain mostly the same code. Could please
> reuse the code using defines in the header file so that the differences
> between the 2 is in the .c file? You might want to look at what
> md32_common.h is doing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openssl/openssl/pull/566#issuecomment-192738844>.
>

Take a look at the SSE optimized code, from which this code was probably
derived.  This code looks like this because it looks like the optimized
code I hope to include later.  While there have been complaints about the
optimized code readability, the optimization is some of the best I have
ever seen.  It is beautiful work from an efficiency point of view.

My preference would be to let this code remain similar to the optimized
code, to reduce the effort to understand all four.  The optimized code has
far less room for factorization.

kroeckx commented:
As far as I can see, the SSE version would just be an implementation of the compression function and we should have some way to select that function.  We should probably just add a function pointer in the CTX that points to the right one, but that's something for later to worry about.  I currently don't see why this would prevent you from not having all this duplicate code.

kroeckx commented on crypto/blake2/blake2b.c:
```diff
@@ -0,0 +1,310 @@
+
+    /* The param struct is carefully hand packed, and should be 64 bytes on
+     * every platform. */
+    OPENSSL_assert(sizeof(BLAKE2B_PARAM) == 64);
```
Can you turn this into a compile time rather than a runtime error?

waywardgeek replied:
On Sun, Mar 6, 2016 at 8:37 AM, Kurt Roeckx <notifications@github.com>
wrote:

> In crypto/blake2/blake2b.c
> <https://github.com/openssl/openssl/pull/566#discussion_r55143610>:
>
> > +    /* The param struct is carefully hand packed, and should be 64 bytes on
> > +     * every platform. */
> > +    OPENSSL_assert(sizeof(BLAKE2B_PARAM) == 64);
>
> Can you turn this into a compile time rather than a runtime error?
>
I am open to suggestions, but I could not find a way to do so that is not
ugly.  For example, I could declare an array:

    size_t dummy[1-(sizeof(BLAKE2B_PARAM) - 64)*(sizeof(BLAKE2B_PARAM) -
64)]

And then actually use it so this does not generate an unused variable
warning.  This gives a compiler error due to 0 or negative array size at
compile time.  This could probably be done in a more readable way with some
sort of pragma, but only weird compilers are going to pack a structure of
64 byte variables in in a non-64 byte memory layout, and I would not expect
pragmas to work in those compilers.

I checked the generated assembly code.  The assertion is optimized away, so
this is not a speed issue.

kroeckx replied:
Just something like:

    #if sizeof(BLAKE2B_PARAM) != 64
    #error "sizeof(BLAKE2B_PARAM) != 64"
    #endif

waywardgeek replied:
On Mon, Mar 7, 2016 at 10:53 AM, Kurt Roeckx <notifications@github.com>
wrote:

> In crypto/blake2/blake2b.c
> <https://github.com/openssl/openssl/pull/566#discussion_r55251662>:
>
> > +
> > +    for(i = 0; i < 8; ++i) {
> > +        S->h[i] = blake2b_IV[i];
> > +    }
> > +}
> > +
> > +/* init xors IV with input parameter block */
> > +static int blake2b_init_param(BLAKE2B_CTX *S, const BLAKE2B_PARAM *P)
> > +{
> > +    size_t i;
> > +    const uint8_t *p = (const uint8_t *)(P);
> > +    blake2b_init0(S);
> > +
> > +    /* The param struct is carefully hand packed, and should be 64 bytes on
> > +     * every platform. */
> > +    OPENSSL_assert(sizeof(BLAKE2B_PARAM) == 64);
>
> Just something like:
>
> #if sizeof(BLAKE2B_PARAM) != 64
> #error "sizeof(BLAKE2B_PARAM) != 64"
> #endif
>
>
I wish that were legal, but sizeof is not allowed in a #if condition.  This
makes sense to me since the actual structure padding is a decision made
late in the compilation process while #if statements happen when still
dealing with input tokens.

I was wrong above when I said that changing the > to >= would cause a
different result.  I tried it, and you were right.  I'm looking into
shrinking the buffer size, which was on of your suggested fixes.

kroeckx replied:
Of course that doesn't work.  So I found those instead:
https://scaryreasoner.wordpress.com/2009/02/28/checking-sizeof-at-compile-time/
http://www.pixelbeat.org/programming/gcc/static_assert.html

Maybe we should just leave it as it is.

waywardgeek replied:
On Mon, Mar 7, 2016 at 2:25 PM, Kurt Roeckx <notifications@github.com>
wrote:

> In crypto/blake2/blake2b.c
> <https://github.com/openssl/openssl/pull/566#discussion_r55283946>:
>
> > +
> > +    for(i = 0; i < 8; ++i) {
> > +        S->h[i] = blake2b_IV[i];
> > +    }
> > +}
> > +
> > +/* init xors IV with input parameter block */
> > +static int blake2b_init_param(BLAKE2B_CTX *S, const BLAKE2B_PARAM *P)
> > +{
> > +    size_t i;
> > +    const uint8_t *p = (const uint8_t *)(P);
> > +    blake2b_init0(S);
> > +
> > +    /* The param struct is carefully hand packed, and should be 64 bytes on
> > +     * every platform. */
> > +    OPENSSL_assert(sizeof(BLAKE2B_PARAM) == 64);
>
> Of course that doesn't work. So I found those instead:
>
> https://scaryreasoner.wordpress.com/2009/02/28/checking-sizeof-at-compile-time/
> http://www.pixelbeat.org/programming/gcc/static_assert.html
>
> Maybe we should just leave it as it is.
>
I would say either leave it as-is, or else let's add an
OPENSSL_compiletime_assert or something like that.

BTW, your suggested change to reduce the buffer size by 2X seems to work!
It simplified the code a bit, made it slightly faster, and saved a bit of
memory.  I have submitted a PR to the BLAKE2 project, and asked if they
could double check this change for mistakes.  However, it passes all the
tests, and I think it just works.  Nice catch.

https://github.com/BLAKE2/BLAKE2/pull/20

richsalz replied:
If we need to, we can add it to to the upcoming platform sanity test.  But for now, let’s leave it.

waywardgeek commented:
On Sat, Mar 5, 2016 at 1:11 PM, Kurt Roeckx <notifications@github.com>
wrote:

> blake2b.c and blake2s.c seem to contain mostly the same code. Could please
> reuse the code using defines in the header file so that the differences
> between the 2 is in the .c file? You might want to look at what
> md32_common.h is doing.
>
I will make this change if you really want me to, but I do not like this
style.  I work on OpenSSL infrequently, and the #define templating used
throughout the code has been the #1 problem for me.  The ASN.1 #define
templates in particular was some of the most cryptic code I have ever read,
and I've read a _lot_ of cryptic code.  Thankfully, that code has been
rewritten.  The small amount of duplicated code (about 60 lines?) between
these two files is easier to maintain and debug separately without the
#define templates, IMO.

waywardgeek commented:
On Sat, Mar 5, 2016 at 5:03 PM, Kurt Roeckx <notifications@github.com>
wrote:

> As far as I can see, the SSE version would just be an implementation of
> the compression function and we should have some way to select that
> function. We should probably just add a function pointer in the CTX that
> points to the right one, but that's something for later to worry about. I
> currently don't see why this would prevent you from not having all this
> duplicate code.
>
There are a lot of optimizations besides in compression.  The Final and
Update functions avoids the temp buffer and byte ordering calls since it
knows that SSE enabled machines are little-endian.

kroeckx commented on crypto/blake2/blake2b.c:
```diff
@@ -0,0 +1,230 @@
+        /* Must be >, not >=, so that last block can be hashed differently */
+        if(datalen > fill) {
+            memcpy(c->buf + c->buflen, in, fill); /* Fill buffer */
+            c->buflen += fill;
```
This line seems useless now.  None of the functions below use it, and then you just set it to 0.

richsalz commented:
@kroeckx , can this go in now and other issues fixed during beta?

waywardgeek commented:
Are there other significant issues remaining?  I tried to either address or
reply to them all.

kroeckx commented:
I'm happy with it.  I've pushed it to the internal review so that others can at least also take a look at it.

kroeckx commented on crypto/evp/m_blake2b.c:
```diff
@@ -0,0 +1,63 @@
+ *
+ * - OpenSSL license   : https://www.openssl.org/source/license.html
+ * - Apache 2.0        : http://www.apache.org/licenses/LICENSE-2.0
+ * - CC0 1.0 Universal : http://www.apache.org/licenses/LICENSE-2.0
```
So here, in m_blake2s.c and in blake2_locl.h the link to the CC0 license is still wrong.  There is no need to fix it here, but maybe other places also still have it wrong.

ekasper replied:
No custom license headers, please!

ekasper replied:
And to clarify (I explained to Kurt internally as well), the top license header MUST be our standard header. But it's fine to add a custom note afterwards, for example

```
/*
 * Derived from the BLAKE2 reference implementation written by Samuel Neves.
 * More information about the BLAKE2 hash function and its implementations
 * can be found at https://blake2.net.
 */
```

waywardgeek replied:
On Wed, Mar 9, 2016 at 3:50 PM, ekasper <notifications@github.com> wrote:

> In crypto/evp/m_blake2b.c
> <https://github.com/openssl/openssl/pull/566#discussion_r55611735>:
>
> > @@ -0,0 +1,63 @@
> > +/*
> > + * BLAKE2 reference source code package - reference C implementations
> > + *
> > + * Copyright 2012, Samuel Neves <sneves@dei.uc.pt>.
> > + * You may use this under the terms of the CC0, the OpenSSL Licence, or the
> > + * Apache Public License 2.0, at your option.  The terms of these licenses can
> > + * be found at:
> > + *
> > + * - OpenSSL license   : https://www.openssl.org/source/license.html
> > + * - Apache 2.0        : http://www.apache.org/licenses/LICENSE-2.0
> > + * - CC0 1.0 Universal : http://www.apache.org/licenses/LICENSE-2.0
>
> And to clarify (I explained to Kurt internally as well), the top license
> header MUST be our standard header. But it's fine to add a custom note
> afterwards, for example
>
> /*
>  * Derived from the BLAKE2 reference implementation written by Samuel Neves.
>  * More information about the BLAKE2 hash function and its implementations
>  * can be found at https://blake2.net.
>  */
>
> I can put the standard OpenSSL copyright header on top, with this comment
> below.  It says:
>
I'm happy to add this at the top:

/* ====================================================================
 * Copyright (c) 2015 The OpenSSL Project. All rights reserved.
 *
 * Rights for redistribution and usage in source and binary
 * forms are granted according to the OpenSSL license.
 */

However, is this technically correct, since Samuel Neves wrote it?  He is
willing to give away the code under any conditions we like, so long as it
does not restrict others in any way to also use it.  I don't think he is
giving the code to the OpenSSL project, just granting a license to the
world to use it under 3 different licenses.

ekasper commented on apps/progs.h:
```diff
@@ -225,6 +225,12 @@ static FUNCTION functions[] = {
 #ifndef OPENSSL_NO_RMD160
     { FT_md, "rmd160", dgst_main},
 #endif
+#ifndef OPENSSL_NO_BLAKE2B
```
Do we really need three different ifdefs for Blake?

levitte replied:
`OPENSSL_NO_BLAKE2B` and `OPENSSL_NO_BLAKE2S` will never be defined as far as I can see, so it's not just unnecessary, these conditions are faulty

ekasper replied:
Please fix, and please ensure this builds with OPENSSL_NO_BLAKE2.

ekasper commented on crypto/blake2/blake2b.c:
```diff
@@ -0,0 +1,226 @@
+
+#include <stdint.h>
+#include <string.h>
+#include <stdio.h>
```
Why does this need stdio.h?

ekasper commented on crypto/blake2/blake2b.c:
```diff
@@ -0,0 +1,226 @@
+
+/* crypto/blake2/blake2b.c */
+
+#include <stdint.h>
```
Our `uint64_t` and friends come from e_os2.h

ekasper commented on crypto/blake2/blake2b.c:
```diff
@@ -0,0 +1,226 @@
+};
+
+/* Some helper functions, not necessarily useful */
+static inline void blake2b_set_lastblock(BLAKE2B_CTX *S)
```
inline -> ossl_inline

ekasper commented on include/openssl/objects.h:
```diff
@@ -642,6 +642,16 @@
+#  define OBJ_blake2b                     1,3,6,1,4,1,1722,12,2,1,16
+
+#  define SN_blake2s                      "BLAKE2s"
+#  define LN_blake2s                      "blake2"
```
blake2 -> blake2s

What happened here? These are supposed to be autogenerated.

kroeckx replied:
I'm not sure this is auto generated?  At least make update doesn't seem to re-add it.

kroeckx replied:
And as far as I can see, the current things seem to follow what the rest of the code is doing.

ekasper replied:
Steve says they are autogenerated from objects.txt. Maybe during 'make' not 'make update' (strangely)?

This one says "blake2" in one file and "blake2s" in another, which looks off.

levitte replied:
Yeah, the missing `s` is tell tale of a manual update.

`make update` *should* have updated it.  If not, then there's a fault in the update procedure...  How about undoing the changes in this file and running `make update` to see?

levitte replied:
(we should also fix the script that does the update to write a note about automatic generation at the top of this file....)

waywardgeek replied:
On Wed, Mar 9, 2016 at 3:42 PM, Kurt Roeckx <notifications@github.com>
wrote:

> In include/openssl/objects.h
> <https://github.com/openssl/openssl/pull/566#discussion_r55610855>:
>
> > @@ -642,6 +642,16 @@
> >  #  define NID_ripemd160WithRSA            119
> >  #  define OBJ_ripemd160WithRSA            1L,3L,36L,3L,3L,1L,2L
> >
> > +#  define SN_blake2b                      "BLAKE2b"
> > +#  define LN_blake2b                      "blake2b"
> > +#  define NID_blake2b                     1022
> > +#  define OBJ_blake2b                     1,3,6,1,4,1,1722,12,2,1,16
> > +
> > +#  define SN_blake2s                      "BLAKE2s"
> > +#  define LN_blake2s                      "blake2"
>
> I'm not sure this is auto generated? At least make update doesn't seem to
> re-add it.
>
I think I wrote those lines by hand when trying to figure out the object
system in OpenSSL.  I removed them in my local version, and it seems to
work without them.  I don't think objects.h gets generated anymore.

waywardgeek replied:
I don't know if my copy is still being tracked, but I committed fixes to
issues raised above, including removing the unused rotl functions, fixing
the NO_BLAKE2 defines (required a minor change to progs.pl), changing
inline to ossl_inline, improving some comments, fixing CC0 links, and
removing blakd2 from objects.h

ekasper commented:
Please also add a CHANGES entry.

ekasper commented on crypto/blake2/blake2_impl.h:
```diff
@@ -0,0 +1,145 @@
+    *p++ = (uint8_t)w;
+}
+
+static inline uint32_t rotl32(const uint32_t w, const unsigned c)
```
I'm squinting hard... are these used?

kroeckx replied:
Only the right ones are used, the left ones not.

ekasper replied:
Then please remove them.

ekasper commented on crypto/blake2/blake2s.c:
```diff
@@ -0,0 +1,221 @@
+    { 10,  2,  8,  4,  7,  6,  1,  5, 15, 11,  9, 14,  3, 12, 13 , 0 } ,
+};
+
+/* Some helper functions, not necessarily useful */
```
Why do we have useless helper functions? :)

kroeckx replied:
It's actually used.

ekasper replied:
Then remove `not necessarily useful` :)

ekasper commented on crypto/blake2/blake2s.c:
```diff
@@ -0,0 +1,221 @@
+    store48(&P->node_offset, 0);
+    P->node_depth    = 0;
+    P->inner_length  = 0;
+    /* memset(P->reserved, 0, sizeof(P->reserved)); */
```
Could you add a comment explaining why this is commented out?

kroeckx replied:
The code on which this is based at least seems to have this too, with the b version having it, and the s version not having it.  The s version just doesn't have a reserved field.

waywardgeek replied:
On Wed, Mar 9, 2016 at 3:47 PM, Kurt Roeckx <notifications@github.com>
wrote:

> In crypto/blake2/blake2s.c
> <https://github.com/openssl/openssl/pull/566#discussion_r55611380>:
>
> > +}
> > +
> > +/* Initialize the hashing context.  Always returns 1. */
> > +int BLAKE2s_Init(BLAKE2S_CTX *c)
> > +{
> > +    BLAKE2S_PARAM P[1];
> > +
> > +    P->digest_length = BLAKE2S_DIGEST_LENGTH;
> > +    P->key_length    = 0;
> > +    P->fanout        = 1;
> > +    P->depth         = 1;
> > +    store32(&P->leaf_length, 0);
> > +    store48(&P->node_offset, 0);
> > +    P->node_depth    = 0;
> > +    P->inner_length  = 0;
> > +    /* memset(P->reserved, 0, sizeof(P->reserved)); */
>
> The code on which this is based at least seems to have this too, with the
> b version having it, and the s version not having it. The s version just
> doesn't have a reserved field.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openssl/openssl/pull/566/files#r55611380>.
>

I just removed it, as well as the commented out reserved bytes in
blake2_locl.h.

ekasper commented on crypto/blake2/blake2s.c:
```diff
@@ -0,0 +1,221 @@
+}
+
+/*
+ * Finalize the hash state in a way that avoids length extension attacks.
```
I'm afraid I don't understand this comment. Why do you talk about length extension here?

kroeckx replied:
In case of the last compress call, it first calls set_lastblock().  I think that's the point of the comment.

ekasper replied:
Then how about something like

```
/*
 * The last compression call of BLAKE2 is different by design,
 * in order to avoid length extension attacks.
 */

(Make it clear that the comment is about the design and not our implementation.)

waywardgeek replied:
On Wed, Mar 9, 2016 at 3:22 PM, ekasper <notifications@github.com> wrote:

> > +/*
>
> > + * Finalize the hash state in a way that avoids length extension attacks.
>
> I'm afraid I don't understand this comment. Why do you talk about length
> extension here?
>
I was asked to add a short comment to every function.  The only interesting
thing about this particular Finalize I see is that unlike SHA256, this
finalize hashes the last block differently, making it impossible to do a
length extension attack like we can do against SHA256.  BLAKE2b/BLAKE2s can
be used to generate MACs directly, without needing HMAC, but the keyed
interface available in BLAKE2 and Keccak are not currently supported in
EVP.

ekasper commented on test/evptests.txt:
```diff
@@ -2,6 +2,63 @@
 #aadcipher:key:iv:plaintext:ciphertext:aad:tag:0/1(decrypt/encrypt)
 #digest:::input:output
 
+# BLAKE2 tests, using same inputs as MD5
```
How were these generated? Are these official / do we have any official test vectors?

kroeckx replied:
They don't seem to match what's in the source, they all seem to also have a key which we don't have.  RFC7693 only seems to have 1 example for each.

kroeckx replied:
RFC7693 also contains a selftest example.

ekasper replied:
Then can we please add the known ones to ensure this actually interoperates?

kroeckx replied:
So the "616263" test that we have, is the "abc" from the RFC, and the result matches.

ekasper replied:
Then add a comment above that test pointing to the RFC! :) It is useful to know where test vectors come from.

waywardgeek replied:
On Wed, Mar 9, 2016 at 3:24 PM, ekasper <notifications@github.com> wrote:

> In test/evptests.txt
> <https://github.com/openssl/openssl/pull/566#discussion_r55609025>:
>
> > @@ -2,6 +2,63 @@
> >  #aadcipher:key:iv:plaintext:ciphertext:aad:tag:0/1(decrypt/encrypt)
> >  #digest:::input:output
> >
> > +# BLAKE2 tests, using same inputs as MD5
>
> How were these generated? Are these official / do we have any official
> test vectors?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openssl/openssl/pull/566/files#r55609025>.
>
I generated them with the official reference source.  All of the official
test vectors are done with key != 0, and since we always set key == 0,
there are no applicable official vectors.

kroeckx commented:
So I pushed some changes to our internal review for this already.

kroeckx commented:
So I created #844 with the changes I've made so far.

richsalz commented:
merged!
